numpy/lib/function_base.py

```diff
@@ -356,7 +356,7 @@ def jhat(nbins):
         p_k = np.histogram(x, bins=nbins)[0] / n
         return (2 - (n + 1) * p_k.dot(p_k)) / hh
 
-    nbins = min(range(1, max(n // 4, 2)), key=jhat)
+    nbins = min(range(1, max(100, int(np.sqrt(n)))), key=jhat)
```

Comment:
I don't know of any strict upper bounds.

This paper uses an upper bound of `5 * n ** (1/3)` (for a different rule), which might be a reasonable choice:
https://arxiv.org/pdf/physics/0605197.pdf